File: bots/intgrah/v54.7.2/builder/tasks/explore.py
"""Walk toward unexplored tiles to grow the bot's known map. Gated on
`ti > EXPLORE_MIN_TI`: exploring lays roads, so a starving bot would
strand titanium it can't recoup. Delegates the actual movement to
`builder.explore`.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Final, override

from builder.explore import explore as run_explore
from builder.tasks.rejected import TaskRejectedError

if TYPE_CHECKING:
    from cambc import Controller

    from builder import Builder

EXPLORE_MIN_TI: Final = 100


class TaskRejectedInsufficientTiError(TaskRejectedError):
    def __init__(self, have: int) -> None:
        self.have = have

    @override
    def __str__(self) -> str:
        return f"ti={self.have} <= {EXPLORE_MIN_TI}; exploring would burn roads we can't recoup"


def explore(self: Builder, ct: Controller) -> None:
    if self.ti <= EXPLORE_MIN_TI:
        raise TaskRejectedInsufficientTiError(self.ti)
    run_explore(self, ct)

File: bots/rs/v56/src/builder/tasks/shared/explore.rs
//! Walk toward unexplored tiles to grow the bot's known map. Gated on
//! `ti > EXPLORE_MIN_TI`: exploring lays roads, so a starving bot would
//! strand titanium it can't recoup. Delegates the actual movement to
//! `builder::explore`.

use cambc::Controller;

use crate::builder::Builder;
use crate::builder::explore::explore as run_explore;
use crate::builder::tasks::rejected::{TaskRejected, TaskResult};

#[pyrust::inline]
const EXPLORE_MIN_TI: i32 = 100;

pub fn explore(self_: &mut Builder, ct: &mut Controller<'_>) -> TaskResult {
    if self_.ti <= EXPLORE_MIN_TI {
        return Some(TaskRejected::from_string(format!(
            "ti={} <= {}; exploring would burn roads we can't recoup",
            self_.ti, EXPLORE_MIN_TI
        )));
    }
    run_explore(self_, ct);
    None
}
